nimble/drivers/nrf5x/src/ble_phy.c

```diff
@@ -1024,26 +1022,168 @@ ble_phy_rx_xcvr_setup(void)
                          RADIO_INTENSET_DISABLED_Msk);
 }
 
+static uint32_t
+ble_phy_transition_anchor_get(uint8_t tifs_anchor, uint8_t phy_state, uint8_t phy_mode)
+{
+    uint32_t time;
+
+    if (tifs_anchor == PHY_TRANS_ANCHOR_END) {
+        /* TX end timestamp is captured in CC[2] */
+        time = NRF_TIMER0->CC[2];
+
+        /* Adjust for delay between EVENT_END and actual TX/RX end time */
+        time += (phy_state == BLE_PHY_STATE_TX)
+                ? g_ble_phy_t_txenddelay[phy_mode]
+                : -g_ble_phy_t_rxenddelay[phy_mode];
+
+    } else {
+        /* RX end timestamp is captured in CC[2] */
+        time = NRF_TIMER0->CC[1];
+
+        /* Adjust for delay between EVENT_ADDRESS and actual AA time ota */
+        time += (phy_state == BLE_PHY_STATE_TX)
+                ? g_ble_phy_t_txaddrdelay[phy_mode]
+                : -g_ble_phy_t_rxaddrdelay[phy_mode];
+
+        /* Adjust by sync word length to get TX/RX start time */
+        time -= ble_ll_pdu_syncword_us(phy_mode);
+    }
+
+    return time;
+}
+
+static void
+ble_phy_transition_fem(uint8_t transition, uint32_t anchor_time)
+{
+#if PHY_USE_FEM
+    /* Note: CC[2] is used for end timestamp which we do not need here. */
+    if (transition == PHY_TRANS_TO_TX) {
+#if PHY_USE_FEM_PA
+        fem_time = anchor_time - MYNEWT_VAL(BLE_FEM_PA_TURN_ON_US);
+        nrf_timer_cc_set(NRF_TIMER0, 2, fem_time);
+        NRF_TIMER0->EVENTS_COMPARE[2] = 0;
+        phy_fem_enable_pa();
+#endif
+    } else {
+#if PHY_USE_FEM_LNA
+        fem_time = anchor_time - MYNEWT_VAL(BLE_FEM_LNA_TURN_ON_US);
+        nrf_timer_cc_set(NRF_TIMER0, 2, fem_time);
+        NRF_TIMER0->EVENTS_COMPARE[2] = 0;
+        phy_fem_enable_lna();
+#endif
+    }
+#endif
+}
+
+static int
+ble_phy_transition(uint8_t transition, uint8_t tifs_anchor, uint16_t tifs_usecs,
+                   uint16_t wfr_usecs, uint8_t phy_state, uint8_t cur_phy_mode,
+                   uint8_t next_phy_mode)
+{
+    uint32_t anchor_time;
+    uint32_t radio_time;
+    uint32_t start_time;
+    uint32_t wfr_time;
+    bool is_late;
+
+    if ((phy_state != BLE_PHY_STATE_TX && phy_state != BLE_PHY_STATE_RX) ||
+        transition == PHY_TRANS_NONE) {
+        ble_phy_disable();
+
+        return 1;
+    }
+
+    anchor_time = ble_phy_transition_anchor_get(tifs_anchor, phy_state, cur_phy_mode);
+    start_time = anchor_time + tifs_usecs;
+    radio_time = start_time;
+
+    ble_phy_transition_fem(transition, start_time);
+
+    if (transition == PHY_TRANS_TO_TX) {
```

Comment:
I'd split this into smaller blocks, i.e.

```
if (transition == PHY_TRANS_TO_TX) {
  ble_transition_to_tx(...);
} else if (transition == PHY_TRANS_TO_RX) {
  ble_transition_to_rx(...);
} else {
  ble_transition_to_none(...);
}
```

the `ble_transition_to_...` functions could combine also FEM code so there's no multiple ifs for the same conditions